### ISSUE TITLE
Can we support the `patchOn` operation without creating a new `ClassFileTransformer`?

### ISSUE THREAD

**Issue by dogourd** [raphw/byte-buddy#1471]
I've noticed that there is a `patchOn` operation in the `AgentBuilder`, which primarily uses the `DifferentialMatcher` to determine which classes need to undergo retransformation. However, the current implementation has two steps: 
1. Removing the old ClassFileTransformer
2. Adding the new ClassFileTransformer.     

One issue with this implementation is that when there are multiple ClassFileTransformers in the environment (such as multiple agents or a single agent with different functionalities), we may want to specify their **execution order**. Performing a `patchOn` operation in this case would change the order, as the old ClassFileTransformer would be moved to the end, resulting in a different bytecode execution order than before. This might not meet expectations in certain scenarios where the fixed execution order of the ClassFileTransformers is desired.

Can we support an alternative implementation without adding or removing ClassFileTransformers, but instead just updating their internal matcher to create a new PatchMode that behaves like INCONSISTENT ( Allow for a short period of time where certain inner classes still apply the old transformer after the matcher has been updated )

**Comment by raphw** [comment]
That's a fair point. I can look into that, maybe I can implement a wrapper transformer with an AtomicReference. The downside of this is that it would require a synchronization on every transformation, but as class loaders already lock this should not be a notable issue.

**Comment by raphw** [comment]
Can you have a look at the latest master and the `PatchMode.SUBSTITUTE` strategy? This should do what you hoped it would.

**Comment by dogourd** [comment]
Sorry for the late feedback.

I did some local testing, but unfortunately the current changes did not work in my test scenarios. The `patchOn` call did not produce any modifications.

I traced the source code a bit, and it seems the problem is likely from `net.bytebuddy.agent.builder.ResettableClassFileTransformer.WithDelegation.Substitutable#substitute`. This method directly modifies the `classfileTransformer` reference held internally by `Substitutable`.

Also, when we call `patchOn`, a `Transformation.DifferentialMatcher` instance is created and holds a reference to the decorated old ClassfileTransformer, which is the `Substitutable`.

The direct reference modification in `substitute` causes the reference held in `DifferentialMatcher` to also change, so ultimately DifferentialMatcher thinks there is no change for this `patchOn` call.

**Comment by raphw** [comment]
Indeed. I'll need to fix the matching, too.

**Comment by dogourd** [comment]
When using the `Substitutable` PatchMode, can users pass in a custom `DifferentialMatcher`? 

Currently in the `patchOn` operation, the `DifferentialMatcher` is used to decide which classes should be retransformed. Of course, the default implementation can ensure that all classes cared about by the updated ClassFileTransformer will be handled properly, but it also has some minor issues - it may retransform many unnecessary classes as well. Many classes we already transformed before the patch operation may not need to be retransformed. We only need to focus on the incremental changes and classes we expect to modify. Users should be fully capable of calculating the incremental changes themselves and deciding which classes to retransform. 

This can minimize the `retransformClasses` behavior and avoid unnecessary class retransformation. Otherwise, based on the current default `DifferentialMatcher` implementation, a lot of classes will be redefined even if the user already knows those classes have no changes.   
Of course, we can work around some classes by controlling the result of `Transformer.equals()` together with the internal `DifferentialMatcher` implementation, but that would create a lot of unnecessary Transformer instances. In the worst case we may need to allocate one Transformer **per class**. 

Minimizing `retransformClasses` is beneficial - the JVM has more restrictions on retransform, and when multiple javaagents are present, it can effectively avoid some conflicts. Some javaagents may even define fields with different names on each retransform call, eventually causing the transform to fail. More seriously, on some early JVM versions, failing to retransform the same class multiple times could even crash the JVM. I saw this issue(#1490 ) which is highly likely to point to this problem: [JDK-8178870 instrumentation.retransformClasses cause coredump](https://bugs.openjdk.org/browse/JDK-8178870).
